cabby/geo/walk.py

```diff
@@ -440,11 +447,63 @@ def get_points_and_route(map: map_structure.Map) -> Optional[item.RVSPath]:
                                                             main_pivot,
                                                             near_pivot,
                                                             beyond_pivot,
-                                                            cardinal_direction)
-
+                                                            cardinal_direction,
+                                                            number_intersections)
     return rvs_path_entity
 
 
+def get_number_interesctions_past(main_pivot: Dict, route: GeoDataFrame, map: map_structure.Map, end_point: Point) -> int:
+    '''Return the number of intersections between the main_pivot and goal. If the main_pivot and goal are on different streets return -1.
+    Arguments:
+      main_pivot: The pivot along the route.
+      route: The route along which a landmark will be chosen.
+      map: The map of a specific region.
+      end_point: The goal location.
+    Returns:
+      The number of intersections between the main_pivot and goal. If the main_pivot and goal are on different streets return -1.
+    '''
+    dist = route['geometry'].apply(
+        lambda x: x.distance(main_pivot['geometry']))
+    pivot_idx = dist.idxmin().tolist()
+    pivot_goal_route = route.loc[pivot_idx:]
+
+    if pivot_goal_route.shape[0] <= 1:
+        return -1
+
+    # Check if main pivot is in the pivot_goal_route segment.
+    segment_pivot = LineString(pivot_goal_route['geometry'].iloc[0:2].values)
+    number_intersection = - \
+        util.project_point_in_segment(
+            segment_pivot, main_pivot['centroid'])
+
+    # Check if main end point is in the pivot_goal_route segment.
+    segment_goal = LineString(pivot_goal_route['geometry'].iloc[-2:].values)
+    number_intersection -= util.project_point_in_segment(
+        segment_goal, end_point['centroid'])
+
+    try:
+        # Check if goal and pivot are on the same street.
+        edges_in_pivot_goal_route = pivot_goal_route['osmid'].apply(
+            lambda x: set(map.edges[map.edges['u'] == x]['osmid'].tolist()))
+
+        pivot_streets = edges_in_pivot_goal_route.iloc[0]
+        goal_streets = edges_in_pivot_goal_route.iloc[-1]
+        common_streets = pivot_streets & goal_streets
+        if not common_streets:
+            return -1
+
+        number_intersection += edges_in_pivot_goal_route.apply(
+            lambda x: len(x - common_streets) > 0).count()
+
+        if number_intersection <= 0:
+            return -1
+    except Exception as e:
+        print(e)
+        return -1
```

Comment:
As before, localize the exception handling and handle explicit named exceptions, please!